sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -159,44 +159,142 @@ class AstBuilder extends DataTypeAstBuilder
   }
 
-  private def visitCompoundBodyImpl(
-      ctx: CompoundBodyContext,
-      label: Option[String],
-      allowVarDeclare: Boolean,
-      labelCtx: SqlScriptingLabelContext,
-      isScope: Boolean): CompoundBody = {
-    val buff = ListBuffer[CompoundPlanStatement]()
-    ctx.compoundStatements.forEach(
-      compoundStatement => buff += visitCompoundStatementImpl(compoundStatement, labelCtx))
+  private def assertSqlState(sqlState: String): Unit = {
+    val sqlStateRegex = "^[A-Za-z0-9]{5}$".r
+    if (sqlStateRegex.findFirstIn(sqlState).isEmpty
+      || sqlState.startsWith("00")
+      || sqlState.startsWith("01")
+      || sqlState.startsWith("XX")) {
+      throw SqlScriptingErrors.invalidSqlStateValue(CurrentOrigin.get, sqlState)
+    }
+  }
+
+  private def visitConditionValueImpl(
+      ctx: ConditionValueContext,
+      handlerTriggers: HandlerTriggers): Unit = {
+    // Current element is SQLSTATE.
+    Option(ctx.sqlStateValue())
+      .foreach { sqlStateValueContext =>
+        val sqlState = sqlStateValueContext.getText.replace("'", "")
+        assertSqlState(sqlState)
+        handlerTriggers.addUniqueSqlState(sqlState)
+      }
 
-    val compoundStatements = buff.toList
+    // Current element is condition.
+    Option(ctx.multipartIdentifier())
+      .foreach { conditionContext =>
+        handlerTriggers.addUniqueCondition(conditionContext.getText)
+      }
 
-    val candidates = if (allowVarDeclare) {
-      compoundStatements.dropWhile {
-        case SingleStatement(_: CreateVariable) => true
-        case _ => false
+    Option(ctx.SQLEXCEPTION())
+      .foreach { _ =>
+        handlerTriggers.addUniqueSqlException()
       }
-    } else {
-      compoundStatements
+
+    // It is sufficient to check only NOT for NOT FOUND handler.
+    Option(ctx.NOT()).foreach { _ =>
+      handlerTriggers.addUniqueNotFound()
     }
+  }
 
-    val declareVarStatement = candidates.collectFirst {
-      case SingleStatement(c: CreateVariable) => c
+  /**
+   * Visit list of condition/sqlstate values in handler declaration.
+   */
+  private def visitConditionValuesImpl(
+      ctx: ConditionValuesContext): HandlerTriggers = {
+
+    val handlerTriggers: HandlerTriggers = new HandlerTriggers()
+
+    ctx.cvList.forEach { cvContext =>
+      visitConditionValueImpl(cvContext, handlerTriggers)
     }
 
-    declareVarStatement match {
-      case Some(c: CreateVariable) =>
-        if (allowVarDeclare) {
-          throw SqlScriptingErrors.variableDeclarationOnlyAtBeginning(
-            c.origin, c.name.asInstanceOf[UnresolvedIdentifier].nameParts)
-        } else {
-          throw SqlScriptingErrors.variableDeclarationNotAllowedInScope(
-            c.origin, c.name.asInstanceOf[UnresolvedIdentifier].nameParts)
+    handlerTriggers
+  }
+
+  private def visitDeclareConditionStatementImpl(
+      ctx: DeclareConditionStatementContext): ErrorCondition = {
+    val conditionName = ctx.multipartIdentifier().getText
```

Comment:
Can we avoid calling `getText`? I'm not sure if it may include backticks or not. Let's use `visitMultipartIdentifier`.